source/s3_auto_ranged_put.c

```diff
@@ -73,11 +85,25 @@ static bool s_process_part_info(const struct aws_s3_part_info *info, void *user_
             break;
     }
 
+    /* Parts might be out of order or have gaps in them. array list allows
+     * setting elements outside of current length, but values between old length
+     * and new value will be uninitialized. as a workaround when processing
+     * element make sure to init any new elements to zeroed values. */
+    size_t current_num_parts = aws_array_list_length(&auto_ranged_put->synced_data.etag_list);
+    if (info->part_number > current_num_parts) {
+        struct aws_byte_buf empty_buf = {0};
+        struct aws_string *null_etag = NULL;
+
+        for (size_t i = info->part_number; i > current_num_parts; --i) {
+            aws_array_list_set_at(&auto_ranged_put->encoded_checksum_list, &empty_buf, i - 1);
+            aws_array_list_set_at(&auto_ranged_put->synced_data.etag_list, &null_etag, i - 1);
+        }
```

Comment:
its actually on purpose to avoid underflowing iterator variable in cases where first part was not completed before (i.e. case where array is empty and we receive data for any part other than 0).

Reply:
I like what you did with it 👍